docs/design/proposals/rbd-mirror.md

```diff
@@ -0,0 +1,140 @@
+
+![mirror](https://i.imgur.com/YF5WeCV.png)
+
+RBD images can be asynchronously mirrored between two Ceph clusters.
+
+## Challenges in ceph-csi
+
+* Currently, ceph-csi generates its unique ID for each rbd image and stores the
+  corresponding PV name and the unique ID mapping, and creates the rbd image
+  with the unique ID and return the encoded value which contains all the
+  required information for other operations. As the same rbd image needs to be
+  used in remote OCP cluster and the remote ceph cluster. In ceph-csi, we need
+  to change the logic of storing/generating the unique ID based on the PVC name
+  and namespace name.
+
+* ceph-csi generated journal(omap data) is not mirrored to the remote cluster.
+
+* Promotion and demotion of RBD image.
+
+## Below provides the outline of how this can be implemented in ceph-csi
+
+* A storage-class need to have a parameter to tell that the image needs to be
+  mirrored.
+* The PVC name and namespace need to be passed to the CSI driver as part of
+  volume create request (rbd image and omap keys/value will be derived from
+  these names if its mirrored image.
+* The RBD mirror daemon required setup has to be done prior enabling the image
+  mirroring from ceph-csi.
+* Promotion and Demotion of RBD image.
+
+### Storage-class parameter
+
+we need to add a new parameter to the storage-class to indicate the image need
+to be mirrored upon creation, based on this parameter ceph-csi will decide few
+things like (information need to store and omap for image identification,
+finding rbd image based on PVC name and namespace name in the remote cluster)
+
+The value for the mirroring will be  either one of these "disabled" (default),
+"journal", or "snapshot"
+
+* "disabled" - (default) no need to mirror the rbd image
+* "journal"  - journal-based mirroring. The mounter should be rbd-nbd as the
+  krbd doesnot support it yet.
+* "snapshot" - snapshot based mirroring. It should work with existing krbd
+  (need ceph octopus+)
+
+As we can spicify a pool in storageclass, we can have multiple storageclass
+serving the same pool but bringing different mirroring capabilities.
+
+**Note**: Older PVC or Snapshots cannot be mirrored
+
+More info on rbd mirroring at [1].
+
+### Pass PVC name and namespace name in CreateVolume
+
+If the app and the PVC created in the local cluster as part of recovery the app
+and the PVC will be recreated in the remote cluster, the data present in the
+PVC in remote  cluster should be the same in the local cluster.
+
+As the PVC name and the namespace name should be sent in volume create request,
+cephcsi will check that the rbd image is already present cluster, we cannot
+check the omap data first as omap data maintained by ceph-csi is not mirrored
+to the remote cluster. This is the challenge ceph-csi need to handle)If the
+image is not present(means its a local cluster) we need to generate the
+required omap data and store its information and create the rbd image, the CSI
```

Comment:
'the CSI' - CSI is only an API specification, it can not do anything. What component should "enable mirroring"?